File: prototypes/lab-report-parser/parsers/aqua.py
"""
Deterministic parser for the Aqua / Tentamus 'INFORME DE ENSAYO / TEST REPORT'
crop-residue lab report (Spanish + English variants of the same template).

No LLM, no OCR. Requires a text-layer PDF (all sample files have one).

    detect(pdf) -> bool     # is this the Aqua template?
    parse(path) -> dict     # {template, header, detected_residues, _validation}

The 'detected residues' are the quantified findings in the
'Resumen Resultados Detectados' / 'RESULTS DETECTED' summary block — NOT the
full multi-residue screening panel (hundreds of '<0,01' rows) that follows.
Every detected row carries a '± uncertainty', which is what separates it from
the screening panel.
"""
import re, sys, json
import pdfplumber

# ---- type detection -------------------------------------------------------
def detect(pdf) -> bool:
    meta = pdf.metadata or {}
    producer = (meta.get("Producer", "") + " " + meta.get("Creator", "")).lower()
    if "aqua esolutions" in producer or "aqua ebs" in producer:
        return True
    first = pdf.pages[0].extract_text() or ""            # anchor fallback
    return ("INFORME DE ENSAYO" in first or "TEST REPORT" in first) and \
           ("Nº Muestra / Informe" in first or "LAB-Ref." in first)

# ---- line reconstruction from word coordinates ----------------------------
# pdfplumber's default line grouping is too tight: it splits a label from its
# value when they sit at slightly different y (label top 220.4 vs value 223.3).
# Regroup words into visual rows with a y-tolerance, then join left-to-right.
def _lines(pdf, y_tol=5):
    out = []
    for pg in pdf.pages:
        words = pg.extract_words(use_text_flow=False)
        words.sort(key=lambda w: (w["top"], w["x0"]))
        rows, cur, base = [], [], None
        for w in words:
            if base is None or abs(w["top"] - base) <= y_tol:
                cur.append(w); base = w["top"] if base is None else base
            else:
                rows.append(cur); cur, base = [w], w["top"]
        if cur: rows.append(cur)
        for r in rows:
            out.append(" ".join(w["text"] for w in sorted(r, key=lambda x: x["x0"])))
    return out

def _word_rows(pdf, y_tol=5):
    """Like _lines, but yields (joined_text, [words]) per visual row, per page."""
    for pg in pdf.pages:
        words = pg.extract_words(use_text_flow=False)
        words.sort(key=lambda w: (w["top"], w["x0"]))
        rows, cur, base = [], [], None
        for w in words:
            if base is None or abs(w["top"] - base) <= y_tol:
                cur.append(w); base = w["top"] if base is None else base
            else:
                rows.append(cur); cur, base = [w], w["top"]
        if cur: rows.append(cur)
        for r in rows:
            r.sort(key=lambda x: x["x0"])
            yield (" ".join(w["text"] for w in r), r)

# ---- spanish/english label -> canonical key -------------------------------
HEADER_MAP = {
    # Spanish
    "Código cliente": "client_code",
    "Referencia cliente": "client_reference",
    "Tipo de muestra": "sample_type",
    "Descripción LAB": "lab_description",
    "Peso muestra (Kg)": "weight_kg",
    "Nº Piezas": "pieces",
    "Estado de la muestra": "sample_condition",
    "Punto de toma de muestra": "sampling_point",
    "Lugar toma de muestra": "sampling_point",
    "Fecha / Hora toma de muestra": "sampling_datetime",
    "Fecha / Hora de recepción": "reception_datetime",
    "Fecha inicio ensayo": "analysis_start",
    "Fecha fin ensayo": "analysis_end",
    "Fecha emisión informe": "report_date",
    # English
    "Customer Code": "client_code",
    "Customer Ref.": "client_reference",
    "Sample type": "sample_type",
    "LAB Description": "lab_description",
    "Sample weight (Kg)": "weight_kg",
    "Pieces No.": "pieces",
    "Condition of sample": "sample_condition",
    "Sample location": "sampling_point",
    "Date / Time sampling": "sampling_datetime",
    "Reception date / time": "reception_datetime",
    "Test start date": "analysis_start",
    "Test completion date": "analysis_end",
    "Report date": "report_date",
}
SAMPLE_ID_ANCHORS = {"Nº Muestra / Informe", "LAB-Ref.:"}
CLIENT_ANCHORS = {"Identificación del cliente", "Customer identification"}
DETECTED_ANCHOR = re.compile(r"(resultados\s+detectados|results\s+detected)", re.I)

# The detected-residues table has several column layouts within this template:
#   - fruit/commodity: Name | Result(±Unc) | MRL-EU | MRL-UK | LOQ
#   - soil:            Name | Result(±Unc) | LOQ | Technique      (no MRLs)
#   - Spanish prints "Result ± Uncertainty"; English omits the ± column
#   - MRL cells can hold a phrase ("It doesn't require", "Ver Suma") not a number
# Rather than hard-code x-positions, we read the column header row to discover
# which columns exist and where, then assign each data cell to the nearest one.
# a method-block header (e.g. "UPLC-MS/MS. LAB 1-01-131") ends the summary and
# starts the full screening panel — stop there so we never ingest the panel.
METHOD_HEADER = re.compile(r"LAB\s*\d+-\d+-\d+")
# page furniture to skip if the summary happens to span a page break
SKIP_ROW = re.compile(r"(Nº Muestra|LAB-Ref|Página|Page:|INFORME DE ENSAYO|TEST REPORT)")

def _gap_cells(words, gap):
    """Cluster a row's words left-to-right into cells separated by x-gaps > gap."""
    words = sorted(words, key=lambda w: w["x0"])
    if not words: return []
    cells, cur = [], [words[0]]
    for w in words[1:]:
        if w["x0"] - cur[-1]["x1"] > gap:
            cells.append(cur); cur = [w]
        else:
            cur.append(w)
    cells.append(cur)
    return cells

def _hdr_key(text):
    """Map a header cell's text to a canonical column key."""
    t = text.lower()
    if "parámetro" in t or "parameter" in t: return "name"
    if "result" in t:                        return "result"   # Resultado / Result
    if "lmr" in t or "mrl" in t:
        if "uk" in t: return "mrl_uk"
        if "ue" in t or "eu" in t: return "mrl_eu"
    if re.search(r"\bl[cq]\b", t):           return "loq"      # LC / LQ
    if "técnica" in t or "tecnica" in t or "techni" in t: return "technique"
    return None

def _columns_from_header(words):
    """Discover [{x0, x1, key}] columns (full x-span) from the table header row."""
    cols = []
    for cell in _gap_cells(words, gap=9):
        key = _hdr_key(" ".join(w["text"] for w in cell))
        if key:
            cols.append({"x0": min(w["x0"] for w in cell),
                         "x1": max(w["x1"] for w in cell), "key": key})
    return cols

def _assign(cols, x):
    """Column key for a word at x0=x: the column whose header x-range is nearest
    (0 distance if x is inside it). Robust to right-aligned numbers that start
    just left of their header label, and to wide 'result ± unc' cells."""
    def dist(c):
        return 0 if c["x0"] <= x <= c["x1"] else min(abs(c["x0"] - x), abs(c["x1"] - x))
    return min(cols, key=dist)["key"]

def _num(s):
    """Strict: whole string is one decimal-comma/dot number -> (float|None, qual)."""
    if s is None: return (None, None)
    m = re.match(r"^([<>]?)\s*([\d.]*\d(?:,\d+)?)$", s.strip())
    if not m: return (None, None)
    return (float(m.group(2).replace(".", "").replace(",", ".")), m.group(1) or None)

def _cellnum(s):
    """Loose: first number in a cell that may carry units, e.g. '0,005 mg/kg'."""
    if not s: return (None, None)
    m = re.search(r"([<>]?)\s*(\d[\d.]*(?:,\d+)?)", s)
    if not m: return (None, None)
    return (float(m.group(2).replace(".", "").replace(",", ".")), m.group(1) or None)

def _detected_residues(pdf):
    """Extract the 'detected results' summary table.

    Reads the column header to discover the schema (fruit has MRL columns, soil
    has a Technique column instead), clusters each data row into cells by x-gap,
    and assigns each cell to the nearest header column. Robust to ES/EN, the
    optional ± uncertainty, phrase-valued MRLs, and multi-word analyte names.
    Stops at the first method-block header so the screening panel is not ingested.
    """
    out, cols, in_block = [], None, False
    for text, words in _word_rows(pdf):
        s = text.strip()
        if DETECTED_ANCHOR.search(s):           # start, or "Continuación..." page
            in_block = True; continue           # keep cols across continuations
        if not in_block:
            continue
        if METHOD_HEADER.search(s):            # screening panel begins -> done
            break
        if SKIP_ROW.search(s):
            continue                            # page furniture across a break
        if s.startswith(("Parámetro", "Parameter")):
            if cols is None:                    # lock schema from the FIRST header
                cols = _columns_from_header(words)  # (continuation headers can be
            continue                            #  shifted vs their own data)
        if cols is None:
            continue                            # data before any header — ignore
        # data row: assign each word to the nearest header column, join per column
        rec = {}
        for w in sorted(words, key=lambda w: w["x0"]):
            key = _assign(cols, w["x0"])
            rec[key] = (rec.get(key, "") + " " + w["text"]).strip()

        name = rec.get("name", "").strip()
        result, rqual = _cellnum(rec.get("result", "").split("±")[0])
        if not name or result is None:
            continue                            # not a valid data row
        unc = _cellnum(rec["result"].split("±")[1])[0] if "±" in rec.get("result", "") else None
        eu, uk = _cellnum(rec.get("mrl_eu", ""))[0], _cellnum(rec.get("mrl_uk", ""))[0]
        row = {
            "analyte": name,
            "result_mgkg": result, "result_qualifier": rqual,
            "uncertainty_mgkg": unc,
            "mrl_eu_mgkg": eu, "mrl_uk_mgkg": uk,
            "loq_mgkg": _cellnum(rec.get("loq", ""))[0],
        }
        # preserve non-numeric MRL text (e.g. "It doesn't require", "Ver Suma")
        if eu is None and rec.get("mrl_eu", "").strip(): row["mrl_eu_note"] = rec["mrl_eu"].strip()
        if uk is None and rec.get("mrl_uk", "").strip(): row["mrl_uk_note"] = rec["mrl_uk"].strip()
        if rec.get("technique"): row["technique"] = rec["technique"].strip()
        out.append(row)
    return out

def parse(path):
    pdf = pdfplumber.open(path)
    if not detect(pdf):
        raise ValueError("Not an Aqua-template report")
    lines = _lines(pdf)

    header = {"sample_id": None, "client_name": None}
    for i, ln in enumerate(lines):
        s = ln.strip()
        if header["sample_id"] is None and s in SAMPLE_ID_ANCHORS and i + 1 < len(lines):
            header["sample_id"] = lines[i + 1].strip()
        if header["client_name"] is None and s in CLIENT_ANCHORS and i + 1 < len(lines):
            header["client_name"] = lines[i + 1].strip()
        m = re.match(r"^\s*([^:]+?):\s*(.+?)\s*$", ln)
        if m:
            key = HEADER_MAP.get(m.group(1).strip())
            if key and not header.get(key):
                header[key] = re.sub(r"\s*\(#\)\s*$", "", m.group(2)).strip()
    if header.get("weight_kg"):
        header["weight_kg"] = _num(header["weight_kg"])[0]

    # detected residues (column-based; handles ES/EN + phrase MRLs)
    detected = _detected_residues(pdf)

    errors = []
    if not header.get("sample_id"): errors.append("missing sample_id")
    if not header.get("sample_type"): errors.append("missing sample_type")
    for r in detected:
        # a detected residue must have a numeric result; MRLs are absent for
        # soil and may legitimately be a phrase ("Ver Suma"), so not required.
        if r["result_mgkg"] is None:
            errors.append(f"unparsed result for {r['analyte']}")

    return {
        "template": "aqua-informe-de-ensayo",
        "header": header,
        "detected_residues": detected,
        "_validation": {"ok": not errors, "errors": errors, "detected_count": len(detected)},
    }

if __name__ == "__main__":
    for p in sys.argv[1:]:
        print("=" * 70); print(p.split("/")[-1])
        try:
            print(json.dumps(parse(p), ensure_ascii=False, indent=2))
        except Exception as e:
            print("  ERROR:", e)

File: src/app/lab-shared/parsers/aqua.ts
// Aqua / Tentamus "INFORME DE ENSAYO / TEST REPORT" (ES + EN). Port of aqua.py.
import {
  Doc, ParseResult, Residue, Word, wordRows, lines, page0Text,
  gapCells, numComma, cellNum,
} from './core';

export function detect(doc: Doc): boolean {
  const prod = doc.producer.toLowerCase();
  if (prod.includes('aqua esolutions') || prod.includes('aqua ebs')) return true;
  const first = page0Text(doc);
  return (first.includes('INFORME DE ENSAYO') || first.includes('TEST REPORT'))
    && (first.includes('Nº Muestra / Informe') || first.includes('LAB-Ref.'));
}

const HEADER_MAP: Record<string, string> = {
  // Spanish
  'Código cliente': 'client_code',
  'Referencia cliente': 'client_reference',
  'Tipo de muestra': 'sample_type',
  'Descripción LAB': 'lab_description',
  'Peso muestra (Kg)': 'weight_kg',
  'Nº Piezas': 'pieces',
  'Estado de la muestra': 'sample_condition',
  'Punto de toma de muestra': 'sampling_point',
  'Lugar toma de muestra': 'sampling_point',
  'Fecha / Hora toma de muestra': 'sampling_datetime',
  'Fecha / Hora de recepción': 'reception_datetime',
  'Fecha inicio ensayo': 'analysis_start',
  'Fecha fin ensayo': 'analysis_end',
  'Fecha emisión informe': 'report_date',
  // English
  'Customer Code': 'client_code',
  'Customer Ref.': 'client_reference',
  'Sample type': 'sample_type',
  'LAB Description': 'lab_description',
  'Sample weight (Kg)': 'weight_kg',
  'Pieces No.': 'pieces',
  'Condition of sample': 'sample_condition',
  'Sample location': 'sampling_point',
  'Date / Time sampling': 'sampling_datetime',
  'Reception date / time': 'reception_datetime',
  'Test start date': 'analysis_start',
  'Test completion date': 'analysis_end',
  'Report date': 'report_date',
};
const SAMPLE_ID_ANCHORS = new Set(['Nº Muestra / Informe', 'LAB-Ref.:']);
const CLIENT_ANCHORS = new Set(['Identificación del cliente', 'Customer identification']);
const DETECTED_ANCHOR = /(resultados\s+detectados|results\s+detected)/i;
const METHOD_HEADER = /LAB\s*\d+-\d+-\d+/;
const SKIP_ROW = /(Nº Muestra|LAB-Ref|Página|Page:|INFORME DE ENSAYO|TEST REPORT)/;

interface Col { x0: number; x1: number; key: string }

function hdrKey(text: string): string | null {
  const t = text.toLowerCase();
  if (t.includes('parámetro') || t.includes('parameter')) return 'name';
  if (t.includes('result')) return 'result';
  if (t.includes('lmr') || t.includes('mrl')) {
    if (t.includes('uk')) return 'mrl_uk';
    if (t.includes('ue') || t.includes('eu')) return 'mrl_eu';
  }
  if (/\bl[cq]\b/.test(t)) return 'loq';
  if (t.includes('técnica') || t.includes('tecnica') || t.includes('techni')) return 'technique';
  return null;
}

function columnsFromHeader(words: Word[]): Col[] {
  const cols: Col[] = [];
  for (const cell of gapCells(words, 9)) {
    const key = hdrKey(cell.map((w) => w.text).join(' '));
    if (key) {
      cols.push({
        x0: Math.min(...cell.map((w) => w.x0)),
        x1: Math.max(...cell.map((w) => w.x1)),
        key,
      });
    }
  }
  return cols;
}

function assign(cols: Col[], x: number): string {
  const dist = (c: Col) => (c.x0 <= x && x <= c.x1 ? 0 : Math.min(Math.abs(c.x0 - x), Math.abs(c.x1 - x)));
  return cols.reduce((best, c) => (dist(c) < dist(best) ? c : best)).key;
}

function detectedResidues(doc: Doc): Residue[] {
  const out: Residue[] = [];
  let cols: Col[] | null = null;
  let inBlock = false;
  for (const { text, words } of wordRows(doc)) {
    const s = text.trim();
    if (DETECTED_ANCHOR.test(s)) { inBlock = true; continue; }
    if (!inBlock) continue;
    if (METHOD_HEADER.test(s)) break;
    if (SKIP_ROW.test(s)) continue;
    if (s.startsWith('Parámetro') || s.startsWith('Parameter')) {
      if (cols === null) cols = columnsFromHeader(words);
      continue;
    }
    if (cols === null) continue;

    const rec: Record<string, string> = {};
    for (const w of [...words].sort((a, b) => a.x0 - b.x0)) {
      const key = assign(cols, w.x0);
      rec[key] = ((rec[key] || '') + ' ' + w.text).trim();
    }
    const name = (rec.name || '').trim();
    const [result, rqual] = cellNum((rec.result || '').split('±')[0]);
    if (!name || result === null) continue;
    const unc = (rec.result || '').includes('±') ? cellNum(rec.result.split('±')[1])[0] : null;
    const eu = cellNum(rec.mrl_eu || '')[0];
    const uk = cellNum(rec.mrl_uk || '')[0];
    const row: Residue = {
      analyte: name,
      result_mgkg: result, result_qualifier: rqual,
      uncertainty_mgkg: unc,
      mrl_eu_mgkg: eu, mrl_uk_mgkg: uk,
      loq_mgkg: cellNum(rec.loq || '')[0],
    };
    if (eu === null && (rec.mrl_eu || '').trim()) row.mrl_eu_note = rec.mrl_eu.trim();
    if (uk === null && (rec.mrl_uk || '').trim()) row.mrl_uk_note = rec.mrl_uk.trim();
    if (rec.technique) row.technique = rec.technique.trim();
    out.push(row);
  }
  return out;
}

export function parse(doc: Doc): ParseResult {
  const ls = lines(doc);
  const header: Record<string, string | number | null> = { sample_id: null, client_name: null };

  ls.forEach((ln, i) => {
    const s = ln.trim();
    if (header.sample_id === null && SAMPLE_ID_ANCHORS.has(s) && i + 1 < ls.length) header.sample_id = ls[i + 1].trim();
    if (header.client_name === null && CLIENT_ANCHORS.has(s) && i + 1 < ls.length) header.client_name = ls[i + 1].trim();
    const m = ln.match(/^\s*([^:]+?):\s*(.+?)\s*$/);
    if (m) {
      const key = HEADER_MAP[m[1].trim()];
      if (key && !header[key]) header[key] = m[2].replace(/\s*\(#\)\s*$/, '').trim();
    }
  });
  if (header.weight_kg) header.weight_kg = numComma(String(header.weight_kg))[0];

  const detected = detectedResidues(doc);

  const errors: string[] = [];
  if (!header.sample_id) errors.push('missing sample_id');
  if (!header.sample_type) errors.push('missing sample_type');
  for (const r of detected) if (r.result_mgkg === null) errors.push(`unparsed result for ${r.analyte}`);

  return {
    template: 'aqua-informe-de-ensayo',
    header,
    detected_residues: detected,
    _validation: { ok: errors.length === 0, errors, detected_count: detected.length },
  };
}
